pkg/operator/controller/canary/controller.go

```diff
@@ -0,0 +1,231 @@
+package canary
+
+import (
+	"fmt"
+	"math/rand"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+	"github.com/openshift/cluster-ingress-operator/pkg/manifests"
+
+	"github.com/google/go-cmp/cmp"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "canary_controller"
+)
+
+var (
+	log = logf.Logger.WithName(controllerName)
+)
+
+// New creates the ingress canary controller.
+//
+// The canary controller will watch the Default IngressController, as well as
+// the canary service, deployment, and route resources.
```

Comment:
My thinking was the canary controller should watch the canary resources for any changes, and trigger a reconcile if someone (or something?) were to modify a canary resource. But sounds that won't be necessary since the canary resources are within their own namespace, and I could see someone making the case for cluster admins having control over the resources anyways without fighting the operator.

For phase 2, my thoughts were to run the route test on a regular interval (~1 minutes, perhaps) via a go routine. For now ill drop the watches on all resources except the default ingress controller.

Reply:
Ah, the watch is so the canary controller can revert external changes to the resources— that makes perfect sense, and I don't know why I didn't realize it before.  * grin *.  Feel free to keep the watches, or drop them if you prefer (maybe it makes sense to allow modification of the canary resources for troubleshooting purposes?).

Reply:
> maybe it makes sense to allow modification of the canary resources for troubleshooting purposes?

Let's go with this suspicion for now. :grin: